### ISSUE TITLE
Protection against integer overflows

### ISSUE THREAD

**Issue by tzeitim** [pola-rs/polars#8595]
### Problem description

It sometimes happens in large workflows that integers are downcasted and future computations suffer from integer overflows (non-polars issue I would say). 

I am not sure about the ease of implementation or whether it would be costly computationally but it would be desirable to throw a warning when some integer overflow occurred. 

In the next example I force one integer overflow. 

In this example I insert numeric variable using `.lit()` and increase it's size by multiplying it by increasingly larger integer numbers until it collapses. What's interesting is that when this number has a larger capacity (like the `10000.0`) this doesn't happen. I believe that the class of the first number is inherited downstream.

Would it be reasonable to have a warning for cases like this?

```
import polars as pl
df = pl.DataFrame({'b':500000})
a_number_that_came_from_another_place = 323566
(df
    .with_columns(pl.lit(a_number_that_came_from_another_place).alias('a'))
    .with_columns([

    (10 * pl.col('a')/pl.col('b')).alias('1'),
    (100 * pl.col('a')/pl.col('b')).alias('2'),
    (1000 * pl.col('a')/pl.col('b')).alias('3'),
    (10000 * pl.col('a')/pl.col('b')).alias('4'),
    (100000 * pl.col('a')/pl.col('b')).alias('5'),
    (100000.0 * pl.col('a')/pl.col('b')).alias('5_float'),
    ]

    )
)
shape: (1, 8)
┌────────┬────────┬─────────┬─────────┬─────────┬──────────────┬──────────────┬─────────┐
│ b      ┆ a      ┆ 1       ┆ 2       ┆ 3       ┆ 4            ┆ 5            ┆ 5_float │
│ ---    ┆ ---    ┆ ---     ┆ ---     ┆ ---     ┆ ---          ┆ ---          ┆ ---     │
│ i64    ┆ i32    ┆ f64     ┆ f64     ┆ f64     ┆ f64          ┆ f64          ┆ f64     │
╞════════╪════════╪═════════╪═════════╪═════════╪══════════════╪══════════════╪═════════╡
│ 500000 ┆ 323566 ┆ 6.47132 ┆ 64.7132 ┆ 647.132 ┆ -2118.614592 ┆ -4006.276736 ┆ 64713.2 │
└────────┴────────┴─────────┴─────────┴─────────┴──────────────┴──────────────┴─────────┘
```

**Comment by ritchie46** [comment]
This would be very expensive to do by default in arithmetic. But we could expose `checked` variants. E.g. `checked_add`, `checked_div` etc.

**Comment by alexander-beedie** [comment]
> we could expose checked variants. 

Or a param on the existing methods? (Which would be tidier in terms of API surface, but not as auto-complete friendly... 🤔)

For example:
```python 
pl.col("x").add( pl.col("y"), checked=True )
pl.col("x").add_checked( pl.col("y") )
```

**Comment by ritchie46** [comment]
In rust those methods already exist in std. In python we could use kwargs, but I believe the function would chain more nicely

**Comment by alexander-beedie** [comment]
Gotcha; separate methods is the more discoverable in any case 👍

**Comment by avimallu** [comment]
While this is being discussed, I want to bring up a nasty surprise which may not be obvious to users that need to find the difference between two "count" operations (which are assigned `pl.UInt32`) where integer overflow protection will help greatly:

```py
>>> import polars as pl
>>> df1 = pl.DataFrame({
...     "A": [1, 2, 2, 3, 3, 3]})
>>> df2 = pl.DataFrame({
...     "B": [1, 1, 1, 2, 3, 3]})
>>> 
>>> (
...     df1
...     .groupby("A")
...     .count()
...     .join(df2
...           .groupby("B")
...           .count(), how="left", left_on="A", right_on="B")
...     .with_columns(
...         pl.col("count").sub(pl.col("count_right")).alias("difference"))
... )
shape: (3, 4)
┌─────┬───────┬─────────────┬────────────┐
│ A   ┆ count ┆ count_right ┆ difference │
│ --- ┆ ---   ┆ ---         ┆ ---        │
│ i64 ┆ u32   ┆ u32         ┆ u32        │
╞═════╪═══════╪═════════════╪════════════╡
│ 1   ┆ 1     ┆ 3           ┆ 4294967294 │
│ 2   ┆ 2     ┆ 1           ┆ 1          │
│ 3   ┆ 3     ┆ 2           ┆ 1          │
└─────┴───────┴─────────────┴────────────┘
```

I understand the rationale behind the output being an unsigned integer - a count obviously cannot be negative.

Here's why it's more important in this case: the average user doesn't really *expect* that the output of `count` will be an unsigned integer (fewer will even know what it is, and even more so since it's not documented). In a large dataframe, this overflow will happen so infrequently that a casual glance (a `head` or just `display`) will miss it.

**Comment by theelderbeever** [comment]
At the risk of asking a dumb question.... would compiling polars with overflow protection enabled be a possible solution?

https://doc.rust-lang.org/cargo/reference/profiles.html#overflow-checks

**Comment by aleewen** [comment]
I ran into this issue recently. It would be great if either an exception was raised or casting to a signed Int was automatically applied. I would assume that raising an exception would be more appropriate. This can be really dangerous for allowing incorrect, unseen data to seep through the cracks.

```python
import polars as pl

df = pl.DataFrame(data={"a": [1], "b": [2]}, schema={"a": pl.UInt32, "b": pl.UInt32})
df.select(pl.col("a") - pl.col("b"))

>>> shape: (1, 1)
┌────────────┐
│ a          │
│ ---        │
│ u32        │
╞════════════╡
│ 4294967295 │
└────────────┘
```

**Comment by KronosTheLate** [comment]
TL;DR: Just showing an example of integer overflow in the wild, with some context.

I recently had some issues from integer overflow in constructing a date dimension table. It was quite unintuitive to me that the code that worked in spark (the operation used to produce the `date_iso_math` column below)  produced weird results in polars. 

I ended up not quite understanding the issue here, and used the `date_iso_string` variant. This is definitively a rough edge that spark did not have. It is clearly a feature and not a bug to provide efficient implementation (no checking) for efficient types (using i8 when it makes sense), but just in case there is an easy way to improve the UI I just wanted to share my experience.

In general, I had the idea in my head that python is overflow-safe, and will yell at you if you overflow. I assume others will make similar conciderations. But of course polars has to be better optimized than that, so 100% a user error. 

<details>

<summary>Code that showcases the issue</summary>

```python
import polars as pl
from polars import col
import datetime

start_date = datetime.date(2012, 1, 1)

end_date = datetime.date(datetime.date.today().year + 3, 12, 31)

date_range = [start_date + datetime.timedelta(days=i) for i in range((end_date - start_date).days + 1)]

df_date = pl.DataFrame({"date": date_range})
df_dim_date = df_date.with_columns(
    year = col("date").dt.year(),
    month = col("date").dt.month(),
    day = col("date").dt.day(),
    date_iso_math              = col("date").dt.year() * 10000 + col("date").dt.month()                * 100 + col("date").dt.day(),
    date_iso_math_with_cast    = col("date").dt.year() * 10000 + col("date").dt.month().cast(pl.Int32) * 100 + col("date").dt.day().cast(pl.Int32),
    month_x100 = col("date").dt.month()*100,
    date_iso_string  = col("date").dt.to_string("%Y%m%d").cast(pl.Int32)
).filter(col("date_iso_math") != col("date_iso_string"))
df_dim_date
```

</details>

Output:
![image](https://github.com/user-attachments/assets/414dcb04-f7e8-4167-8066-84192ca78084)

**Comment by KronosTheLate** [comment]
Would it be technically possible to have operations like the `col("date").dt.month() * 100` result in type promotion to whatever type the literal `100` has in python? This would both be reasonable, and remove the rough edge I experienced.

**Comment by cmdlineluser** [comment]
It seems you can also get 3 different result types depending on how the multiplication is done.

```python
import polars as pl

pl.Config(
    tbl_hide_dtype_separator=1, 
    tbl_hide_dataframe_shape=1,
    tbl_hide_column_names=1
)

df = pl.select(x = 12).cast(pl.Int8)

df * 100
# ┌──────┐
# │ i64  │
# ╞══════╡
# │ 1200 │
# └──────┘

df.select(pl.col.x * 100)
# ┌─────┐
# │ i8  │
# ╞═════╡
# │ -80 │
# └─────┘

df.select(pl.col.x * pl.lit(100, int))
# ┌──────┐
# │ i64  │
# ╞══════╡
# │ 1200 │
# └──────┘

df.with_columns(y = 100).select(pl.col.x * pl.col.y)
# ┌──────┐
# │ i32  │
# ╞══════╡
# │ 1200 │
# └──────┘
```
